File: Cpp/1138.字母板上的路径.cpp
/*
 * @lc app=leetcode.cn id=1138 lang=cpp
 *
 * [1138] 字母板上的路径
 */

#include <iostream>
#include <string>

using namespace std;

// @lc code=start
class Solution {
public:
    string alphabetBoardPath(string target) {
        int cx = 0, cy = 0;
        string ans;
        for (char c : target) {
            int nx = (c - 'a') / 5;
            int ny = (c - 'a') % 5;
            if (nx < cx) {
                ans.append(cx - nx, 'U');
            }
            if (ny < cy) {
                ans.append(cy - ny, 'L');
            }
            if (nx > cx) {
                ans.append(nx - cx, 'D');
            }
            if (ny > cy) {
                ans.append(ny - cy, 'R');
            }
            ans.push_back('!');
            cx = nx;
            cy = ny;
        }
        return ans;
    }
};
// @lc code=end

int main(int argc, const char * argv[]) {
    // insert code here...
    
    string target;
    string res;
    
    /**
     输入：target = "leet"
     输出："DDR!UURRR!!DDD!"
     */
    target = "leet";
    res = "DDR!UURRR!!DDD!";
    assert(res == Solution().alphabetBoardPath(target));
    
    /**
     输入：target = "code"
     输出："RR!DDRR!UUL!R!"
     */
    target = "code";
    res = "RR!DDRR!UUL!R!";
    assert(res == Solution().alphabetBoardPath(target));
    

    cout << "OK~" << endl;
    
    return 0;
}

File: Swift/1138.字母板上的路径.swift
/*
 * @lc app=leetcode.cn id=1138 lang=swift
 *
 * [1138] 字母板上的路径
 */

// @lc code=start
class Solution {
    func alphabetBoardPath(_ target: String) -> String {
        var cx = 0, cy = 0
        var ans = ""
        for c in target {
            let nx = Int(c.asciiValue! - 97) / 5
            let ny = Int(c.asciiValue! - 97) % 5
            if nx < cx {
                ans.append(.init(repeating: "U", count: cx - nx))
            }
            if ny < cy {
                ans.append(.init(repeating: "L", count: cy - ny))
            }
            if nx > cx {
                ans.append(.init(repeating: "D", count: nx - cx))
            }
            if ny > cy {
                ans.append(.init(repeating: "R", count: ny - cy))
            }
            ans.append("!")
            cx = nx
            cy = ny
        }
        return ans
    }
}
// @lc code=end

func main() {
    var target: String
    var res: String
    
    /**
     输入：target = "leet"
     输出："DDR!UURRR!!DDD!"
     */
    target = "leet"
    res = "DDR!UURRR!!DDD!"
    assert(res == Solution().alphabetBoardPath(target))
    
    /**
     输入：target = "code"
     输出："RR!DDRR!UUL!R!"
     */
    target = "code"
    res = "RR!DDRR!UUL!R!"
    assert(res == Solution().alphabetBoardPath(target))
}
